### DISCUSSION TITLE
Ошибка 401

### DISCUSSION THREAD

**Discussion by podorojnique** [Chimildic/goofy#208] [Помощь с ошибкой] [answered]
Спотифай сбросил мне пароль из-за подозрительной активности, я его сменил. После этого все скрипты перестали работать. Попробовал дать права доступа - ничего не помогло.
Решил все заново переустановить. Создал пока единственный плейлист Smartify:

function Smartify() {
  let HistoryDay = RecentTracks.get(300); // недавно прослушанные треки
  let SmartifyDay = Source.getPlaylistTracks('Smartify', '0IvahETPKWYzu8h89DjWWJ'); // вчерашний плейлист дня Smartify                    (Создать плейлист и вставить его ID вместо abc)
  Filter.removeTracks(HistoryDay, SmartifyDay, true); // треки Smartify прослушанные вчера
  Cache.append('HistorySmartify', HistoryDay, 'begin'); // добавление вчерашних треков в файл истории Smartify
  let History = Cache.read('HistorySmartify'); // история Smartify из кэша
  Filter.dedupTracks(History); // удаление дубликатов
  Cache.compressTracks(History); //сжатие файла 
  Cache.write('HistorySmartify', History); // запись истори прослушивания Smartify в кэш  
    let Saved = Source.getSavedTracks(); // любимые треки
  Combiner.push(History, Saved); // прослушанное ранее - история плюс любимые треки
    let BigMix = Source.getTracks([
        { name: 'Микс дня 1', id: '37i9dQZF1E3741x9cYnw48' },
        { name: 'Микс дня 2', id: '37i9dQZF1E37b8hVjMv1qr' },
        { name: 'Микс дня 3', id: '37i9dQZF1E37sl1ZyYBLmM' },
        { name: 'Микс дня 4', id: '37i9dQZF1E35FqVXrxpN7h' },
        { name: 'Микс дня 5', id: '37i9dQZF1E3a3uRiEjXzgd' },
        { name: 'Микс дня 6', id: '37i9dQZF1E37y5UA84JDVa' },
  ]); // миксы дня                                                                                                         (Во всех миксах вставить свой ID вместо abc)
  Filter.removeTracks(BigMix, History); // удаление прослушанного ранее
  Order.shuffle(BigMix); // перемешивание
    let Discovery = Source.getPlaylistTracks('Открытия недели', '37i9dQZEVXcVm4EflhNwwv'); // Открытия недели                                              (Вставить ID Открытий вместо abc)
  Filter.removeTracks(Discovery, History); // удаление прослушанного ранее
  let Radar = Source.getPlaylistTracks('Радар новинок', '37i9dQZEVXbnlkegXAWj3s'); // Радар новинок                                                        (Вставить ID Радара вместо abc)
  Filter.removeTracks(Radar, History); // удаление прослушанного ранее
  let Recommendation = Combiner.alternate('max',Discovery , Radar); // поочередное смешивание Радара и Открытий
  let Personally = Combiner.alternate('max', Recommendation, BigMix); // поочередное смешивание новых треков из персональных плейлистов
  let Selected = Source.getFollowedTracks(); // треки из избранных плейлистов     
  Filter.removeTracks(Selected, History); // удаление прослушанного ранее
  Order.shuffle(Selected); // перемешивание
  let Fresh = Source.getFollowedTracks({limit: 10}); // треки из 10 случайных избранных плейлистов                          (Можно изменить 10 на требуемое количество)
  Filter.removeTracks(Fresh, History); // удаление прослушанного ранее
  Filter.rangeDateRel(Fresh, 20); // свежие треки добавленные в последние 20 дней                                                (Можно изменить 20 на требуемое количество)
  Order.shuffle(Fresh); // перемешивание
  
  let Follow = Combiner.alternate('max', Selected, Fresh); // смешивание треков из избранных плейлистов чередуя со свежими
  Filter.dedupTracks(Follow); // удаление дубликатов треков
  let New = Combiner.mixin(Personally , Follow, 1, 1, false); // смешивание новых треков                                                   (Можно изменить соотношение)
  // из персональных и избранных плейлистов в соотношении 1 к 1                                             (если не используются избранные плейлисты  выставить 1, 0 )  

  Filter.dedupArtists(New); // удаление дубликатов исполнителей
  Filter.match(New, 'live', true); // удаление live версий                                                           (Ниже возможно добавление дополнительных фильтров)
  
  let HistoryLast = RecentTracks.get(1000); // последние прослушанные 1000 треков
  Filter.rangeDateRel(HistoryLast, 3); // история прослушивания за последние 3 дня                                         (Можно изменить 3 на требуемое количество !)
  Filter.removeArtists(New, HistoryLast); // удаление треков артистов игравших последние 3 дня
  let SavedTracks = Cache.read('HistorySmartify'); // история прослушивания Smartify
  Filter.removeTracks(SavedTracks, Saved, true); // оставление в истории Smartify только любимых треков
  Selector.keepAllExceptLast(SavedTracks, 100); // отсекание "буфера" любимых треков
  Filter.removeTracks(Saved, SavedTracks); // любимые треки не прослушанные в Smartify
  Filter.removeArtists(Saved, HistoryLast); // удаление треков артистов игравших последние 3 дня
  Order.shuffle(Saved); // перемешивание
  let Shot = Source.getSavedTracks(); // любимые треки
  Filter.rangeDateRel(Shot, 6);  // треки залайканые за последние 6 дней - "контрольный выстрел"                 (Должно быть в два раза больше предыдущего параметра !)
  Filter.removeArtists(Shot, HistoryLast); // удаление треков артистов игравших последние 3 дня
  let Loved = Combiner.alternate('max', Saved, Shot); // поочередное смешивание любимых  треков и "контрольного выстрела"
  Filter.dedupArtists(Loved); // удаление дубликатов исполнителей  

  let Smartify = Combiner.mixin(Loved, New, 1, 4, false); // смешивание Любимых и Новых треков                                              (Можно изменить соотношение)
                                                          //  в соотношении 1 к 1                                  (если не используются любимые треки  выставить 0, 1 )
  Filter.dedupArtists(Smartify); // удаление дубликатов исполнителей
  Selector.keepFirst(Smartify, 100); // установление размера плейлиста Smartify                                   (Можно изменить количество треков итогового плейлиста)
  Playlist.saveWithReplace({id: '0IvahETPKWYzu8h89DjWWJ', name: 'Smartify', tracks: Smartify}); // запись треков в плейлист Smartify            (Вставить ID Плейлиста Smartify вместо abc)                 
}

Добавил все ID которые требовались, триггеры пока не настраивал но при выполнении этого плейлиста появляется ошибка:

Номер: 401
Адрес: https://api.spotify.com/v1/playlists/2MXKwqv9X4l6P8z8cXAHf0/tracks?offset=200&limit=50&locale=RU
Текст ответа: {
  "error" : {
    "status" : 401,
    "message" : "Unauthorized."
  }
} 

Описание и решение ошибок: https://chimildic.github.io/goofy/#/errors

**Comment by Chimildic** [comment, marked as answer]
Ошибка говорит об отсутствии авторизации. Вероятно вы что-то не доделали при обновлении.
Обычно для выдачи доступа хватает [таких](https://chimildic.github.io/goofy/#/tuning?id=Обновить-права-доступа) шагов.

**Reply by podorojnique** [reply]
Спасибо, помогло. Только почему-то на третий раз, два раза перелогинился, давал права доступа - не помогало. На третий раз завелось (:
